Fix disable_sstable_write synchronization with on_compaction_completion

bhalevy opened the pull request:
`disable_sstable_write` needs to acquire `_sstable_deletion_sem` to properly synchronize with background deletions done by `on_compaction_completion` to ensure no sstables will be created or deleted during `reshuffle_sstables` after `storage_service::load_new_sstables` disables sstable writes.

Fixes #4622

Test: unit(dev), nodetool_additional_test.py migration_test.py

raphaelsc commented:
other than my concern above, it looks good to me

bhalevy commented:
> wouldn't this bloat the log with tons of message when LCS is used for example?

I think the overhead is worth it and we're logging anyway when compacting each sstable so it should be comparable.

glommer commented:
One compaction can involve upwards to 32 SSTables, and have it happening from all shards.

Keep in mind that there is a plan to remove even the compaction messages we have today. Avi even merged a patch for that at some point and I begged him to revert it, because the compaction table as we have to day does not have the full info needed for that -> but when it does, the plan is still to remove even the existing messages.

I am totally against moving this message to info.

bhalevy commented:
OK, I'll drop this patch then.

On Wed, Jul 10, 2019, 20:08 Glauber Costa <notifications@github.com> wrote:

> One compaction can involve upwards to 32 SSTables, and have it happening
> from all shards.
>
> Keep in mind that there is a plan to remove even the compaction messages
> we have today. Avi even merged a patch for that at some point and I begged
> him to revert it, because the compaction table as we have to day does not
> have the full info needed for that -> but when it does, the plan is still
> to remove even the existing messages.
>
> I am totally against moving this message to info.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/4691?email_source=notifications&email_token=AE7RGOBHLXQK6DCFDUKGOTLP6YJSVA5CNFSM4H7OFOYKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZUD37I#issuecomment-510148093>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE7RGOBX7QUG7DRNVTFHJP3P6YJSVANCNFSM4H7OFOYA>
> .
>

bhalevy commented:
@glommer
In v2 ([6dad9ba](https://github.com/bhalevy/scylla/commits/6dad9baa1c0e4634f69d698612dd3d5dd7133f1d)):
- dropped `sstable: remove_by_toc_name: log info message`
- rebased and retested unit(dev)

avikivity commented on table.cc:
```diff
@@ -1958,6 +1958,7 @@ future<int64_t>
     _sstable_writes_disabled_at = std::chrono::steady_clock::now();
     return _sstables_lock.write_lock().then([this] {
+      return _sstable_deletion_sem.wait().then([this] {
         if (_sstables->all()->empty()) {
```
Please document the lock order in a follow up.

Perhaps we should fold _sstable_deletion_sem into _sstables_lock.

bhalevy replied:
documented in 0e4567c8815e4bc7f6acc72aee88194e75a2107f

Regarding folding into `_sstables_lock` , they are currently taken at different times in the table::compact_sstables -> on_compaction_completion path, and I'm not sure about the table::snapshot path.  I think that keeping them separate makes sense.

avikivity commented:
@glommer are you done with your review?

glommer commented:
Yes, the only thing unreasonable for me was the info message.

bhalevy commented:
@avikivity can you please merge this?

slivne commented on sstables/sstables.cc:
```diff
@@ -2907,7 +2907,7 @@ remove_by_toc_name(sstring sstable_toc_name, const io_error_handler& error_handl
         sstring dir;
 
-        sstlog.debug("Removing by TOC name: {}", sstable_toc_name);
+        sstlog.info("Removing SSTable {}", sstable_toc_name);
```
@bhalevy did you address this ?

bhalevy replied:
Yes. As I mentioned:

> In v2 ([6dad9ba](https://github.com/bhalevy/scylla/commits/6dad9baa1c0e4634f69d698612dd3d5dd7133f1d)):
> 
> * dropped `sstable: remove_by_toc_name: log info message`
> * rebased and retested unit(dev)

slivne commented:
@avikivity please merge we need this for 3.1

bhalevy commented:
@avikivity please backport to 3.1 and 3.0

avikivity commented:
Backported to 3.1.

avikivity commented:
@bhalevy please prepare a 3.0 backport

bhalevy commented:
@avikivity 3.0 backport here: https://github.com/bhalevy/scylla/commits/fix-disable-sstable-write-for-3.0
based on branch-3.0 d77aaada86885b1a08b90a9a6cda3379cd0735a9

avikivity commented:
thanks, queued as a198db31dc754fcfc1170c7be8acba8e01955a13
